server/app/controllers/applicant/ApplicantProgramsController.java

```diff
@@ -275,14 +275,38 @@ public CompletionStage<Result> editWithApplicantId(
   }
 
   @Secure
-  public CompletionStage<Result> edit(Request request, long programId) {
+  public CompletionStage<Result> edit(
+      Request request, String programParam, Boolean maybeProgramSlug) {
+    // Redirect home when the program slug URL feature is enabled and the program param could be
+    // a program slug but it is actually a program id (numeric).
+    // TODO(#10763): Add metrics to track how often this happens to decide whether we should make
+    // this case invalid or not
+    boolean programSlugUrlEnabled = settingsManifest.getProgramSlugUrlsEnabled(request);
+    if (programSlugUrlEnabled && maybeProgramSlug && StringUtils.isNumeric(programParam)) {
+      return CompletableFuture.completedFuture(redirectToHome());
+    }
+
     Optional<Long> applicantId = getApplicantId(request);
     if (applicantId.isEmpty()) {
       // This route should not have been computed for the user in this case, but they may have
       // gotten the URL from another source.
       return CompletableFuture.completedFuture(redirectToHome());
     }
-    return editWithApplicantId(request, applicantId.orElseThrow(), programId);
```

Comment:
This may have not been the best first route to change, since it just redirects to another one. However, we still need to update it since we are not deprecating any route, yet